Readme.md

```diff
@@ -1,5 +1,298 @@
+
+# Lexer
+
+## What is a Lexer
+- A Lexer is part of an interpreter that takes sequence of characters and turns them into tokens.
+
+## How to use the Lexer
+
+First define your BNF input
+```rust
+let bnf_input = include_str!("path/to/example.bnf");
+```
+
+To parse the input you should use `flux_bnf::bnf::parse`
+```rust
+let mut lexer = flux_bnf::bnf::parse(bnf_input).unwrap();     
+```
+
+## How to set Lexer Rules
+Now that you have `Lexer` we can use that to now set some rules
+
+Before being able to set some rules you will need write a BNF describing the syntax which you can find [here](#bnf)
+
+Here is an example of what our BNF looks like
+```
+root ::= NumberList
+sep ::= " "
+number ::= [0-9]+
+NumberList ::= number (sep number)*
+```
+
+Token input using BNF
+`1 2 3`
+
+Token output
+```
+Some(
+    Token {
+        name: Some(
+            "NumberList",
+        ),
+        match: "1 2 3",
+        range: 0..5,
+        children: [
+            Token {
+                name: Some(
+                    "number",
+                ),
+                match: "1",
+                range: 0..1,
+            },
+            Token {
+                name: Some(
+                    "number",
+                ),
+                match: "2",
+                range: 2..3,
+            },
+            Token {
+                name: Some(
+                    "number",
+                ),
+                match: "3",
+                range: 4..5,
+            },
+        ],
+        ..
+    },
+)
+```
+To apply the rules to the `lexer` you simpily do
+```rust
+lexer.add_rule_for_names(vec!["sep"], CullStrategy::DeleteAll);
+```
+
+This is what the tokenized tree would look like before applying the rules
+```
+Some(
+    Token {
+        name: Some(
+            "NumberList",
+        ),
+        match: "1 2 3",
+        range: 0..5,
+        children: [
+            Token {
+                name: Some(
+                    "number",
+                ),
+                match: "1",
+                range: 0..1,
+            },
+            Token {
+                name: Some(
+                    "sep",
+                ),
+                match: " ",
+                range: 1..2,
+            },
+            Token {
+                name: Some(
+                    "number",
+                ),
+                match: "2",
+                range: 2..3,
+            },
+            Token {
+                name: Some(
+                    "sep",
+                ),
+                match: " ",
+                range: 3..4,
+            },
+            Token {
+                name: Some(
+                    "number",
+                ),
+                match: "3",
+                range: 4..5,
+            },
+        ],
+        ..
+    },
+)
+```
+This is what the tree would look like after the rules were applied. You can see that the seps were taken out of the tree making it easier to read. 
```

Comment:
This explanation should be higher up - explain what is being done and why first, then show the outcome, then summarize again if needed.